### ISSUE TITLE
New gain reference file failing in ramp-fit step

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#6028]
_Issue [JP-2080](https://jira.stsci.edu/browse/JP-2080) was created on JIRA by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert):_

I'm creating some new gain reference files for NIRCam, and the ramp-fitting step is failing when I try running it with them. I'm using GainModel to save the reference files.  I tried running ramp-fitting with the reference file in CRDS and it completed successfully. Looking at the differences between the new reference file and the current one, I see that the new reference file does not have an ASDF extension, while the current one does. The current reference file also has several CHECKSUM/DATASUM keywords that the new file does not.

I will attach my script for creating the reference files. Below is the fitsdiff between the old reference file and the new one, along with the traceback for the ramp-fitting error.

Speaking with James Davies about this he suggested:

There is a switch (skip_fits_update) in datamodels that is turned on by default that allows the data model reader to check if the fits portion of the file has been fiddled with since it was last synced with the ASDF extension. If the checksum is good it doesn’t bother reading from the fits header and just uses the ASDF extension as-is. This should be turned off for reference files that don’t have ASDF extensions. But given that these reference files don’t have ASDF extensions they probably shouldn’t have checks sums either. 

 

```

fitsdiff nircam_nrca1_bootstrapped_gain.fits █████████████████████████████████████████████████████████

 

 fitsdiff: 4.2.1

 a: nircam_nrca1_bootstrapped_gain.fits

 b: █████████████████████████████████████████████████████████

 Maximum number of different data values to be reported: 10

 Relative tolerance: 0.0, Absolute tolerance: 0.0

 

Files contain different numbers of HDUs:

 a: 2

 b: 3

 

Primary HDU:

 

   Headers contain differences:

     Headers have different number of cards:

      a: 28

      b: 68

     Extra keyword 'CERTIFYD' in b: 'PASSED'

     Extra keyword 'CHECKSUM' in b: 'NePlQZNjNbNjNZNj'

     Extra keyword 'DATASUM' in b: '0'

     Extra keyword 'VERIFIED' in b: 'PASSED'

     Inconsistent duplicates of keyword 'HISTORY':

      Occurs 5 time(s) in a, 41 times in (b)

     Keyword DATE     has different values:

        a> 2021-05-14T13:24:45.287

        b> 2018-02-01T19:11:47.021

     Keyword DESCRIP  has different values:

        a> Gain values bootstrapped from CV3 data

        b> Gain refrence file from CV3 data using improved dPTC method

     Keyword DESCRIP  has different comments:

        a> Description of the referenc

        b> Descri

     Keyword FILENAME has different values:

        a> nircam_nrca1_bootstrapped_gain.fits

        b> NIRCam_NRCA1_CV3_128x128_excludingbadfile_gain.fits

     Keyword FILENAME has different comments:

        a> Name of the file

         ?               --

        b> Name of the fi

     Keyword HISTORY  has different values:

        a> This file contains gain values bootstrapped from CV3 flat field data and

        b> This is a NIRCam gain reference file created using an improved method ov

     Keyword HISTORY [2] has different values:

        a>  full well information. Not enough data were collected in CV3 to be able

        b> er the standard photon transfer curve (PTC) method. This method, called

     Keyword HISTORY [3] has different values:

        a>  to produce a high signal to noise gain map across the detector, so we u

        b> differential PTC (dPTC) is less sensitive to residual non-linearity effe

     Keyword HISTORY [4] has different values:

        a> se a single value for all pixels. Calculation of these values was done b

        b> cts, and does not require exposures as fully calibrated as the files nee

     Keyword HISTORY [5] has different values:

        a> y Misselt.

        b> ded for PTC. This leads to more accurate gain value calculations. For mo

     Keyword SUBARRAY has different values:

        a> N/A

        b> GENERIC

 

Extension HDU 1 (SCI, 1):

 

   Headers contain differences:

     Headers have different number of cards:

      a: 9

      b: 11

     Extra keyword 'CHECKSUM' in b: 'ckaDekTAckYAckYA'

     Extra keyword 'DATASUM' in b: '843437942'

 

   Data contains differences:

     Data differs at [1, 1]:

        a> 2.05

        b> 0.0

     Data differs at [2, 1]:

        a> 2.05

        b> 0.0

     Data differs at [3, 1]:

        a> 2.05

        b> 0.0

     Data differs at [4, 1]:

        a> 2.05

        b> 0.0

     Data differs at [5, 1]:

        a> 2.05

        b> 0.0

     Data differs at [6, 1]:

        a> 2.05

        b> 0.0

     Data differs at [7, 1]:

        a> 2.05

        b> 0.0

     Data differs at [8, 1]:

        a> 2.05

        b> 0.0

     Data differs at [9, 1]:

        a> 2.05

        b> 0.0

     Data differs at [10, 1]:

        a> 2.05

        b> 0.0

     ...

     4194304 different pixels found (100.00% different).

```

 

```

2021-05-14 09:25:30,570 - stpipe.RampFitStep - INFO - RampFitStep instance created.

2021-05-14 09:25:33,281 - stpipe.RampFitStep - INFO - RampFitStep instance created.

2021-05-14 09:25:33,338 - stpipe.RampFitStep - INFO - Step RampFitStep running with args ('/Users/hilbert/python_repos/test_mirage_yaml_generator/catalog_generation_test_data/jw42424001001_01101_00004_nrcb5_linear.fits',).

2021-05-14 09:25:33,339 - stpipe.RampFitStep - INFO - Step RampFitStep parameters are: \{'pre_hooks': [], 'post_hooks': [], 'output_file': '/System/Volumes/Data█████████████████████████████████████████████████████████████████████████████████ 'output_dir': None, 'output_ext': '.fits', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': None, 'search_output_file': True, 'input_dir': '', 'int_name': '', 'save_opt': False, 'opt_name': '', 'maximum_cores': 'none'}

2021-05-14 09:25:33,703 - stpipe.RampFitStep - INFO - Using READNOISE reference file: ███████████████████████████████████████████████████████████████

2021-05-14 09:25:33,764 - stpipe.RampFitStep - INFO - Using GAIN reference file: /System/Volumes/Data████████████████████████████████████████████████████████████████████████████████████████████████████████

2021-05-14 09:25:35,568 - stpipe.RampFitStep - INFO - Using algorithm = ols

2021-05-14 09:25:35,568 - stpipe.RampFitStep - INFO - Using weighting = optimal

Traceback (most recent call last):

  File "pipeline_test.py", line 14, in <module>

    output = d.call(linfile, override_gain=gainfile, output_file='slope.fits')

  File "/Users/hilbert/miniconda3/envs/nircam_calib/lib/python3.7/site-packages/stpipe/step.py", line 590, in call

    return instance.run(*args)

  File "/Users/hilbert/miniconda3/envs/nircam_calib/lib/python3.7/site-packages/stpipe/step.py", line 407, in run

    step_result = self.process(*args)

  File "/Users/hilbert/miniconda3/envs/nircam_calib/lib/python3.7/site-packages/jwst/ramp_fitting/ramp_fit_step.py", line 73, in process

    self.weighting, max_cores

  File "/Users/hilbert/miniconda3/envs/nircam_calib/lib/python3.7/site-packages/jwst/ramp_fitting/ramp_fit.py", line 110, in ramp_fit

    gain_2d, weighting, max_cores)

  File "/Users/hilbert/miniconda3/envs/nircam_calib/lib/python3.7/site-packages/jwst/ramp_fitting/ramp_fit.py", line 222, in ols_ramp_fit_multi

    input_model.meta.exposure.drop_frames1, int_times)

  File "/Users/hilbert/miniconda3/envs/nircam_calib/lib/python3.7/site-packages/jwst/ramp_fitting/ramp_fit.py", line 757, in ols_ramp_fit

    f_max_seg)

  File "/Users/hilbert/miniconda3/envs/nircam_calib/lib/python3.7/site-packages/jwst/ramp_fitting/ramp_fit.py", line 1899, in calc_slope

    rn_sect, gain_sect, ngroups, weighting, f_max_seg)

  File "/Users/hilbert/miniconda3/envs/nircam_calib/lib/python3.7/site-packages/jwst/ramp_fitting/ramp_fit.py", line 2194, in fit_next_segment

    sig_intercept, sig_slope, inv_var, save_opt)

  File "/Users/hilbert/miniconda3/envs/nircam_calib/lib/python3.7/site-packages/jwst/ramp_fitting/utils.py", line 181, in append_arr

    self.slope_2d[num_seg[g_pix], g_pix] = slope[g_pix]

IndexError: index 1 is out of bounds for axis 0 with size 1

```

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-2080?focusedCommentId=527187#comment-527187):_

Tagging [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies)

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-2080?focusedCommentId=529634#comment-529634):_

I also tried opening one of the reference files currently in CRDS and modifying the data values within, but that fails when I attempt to save the updated model instance.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2080?focusedCommentId=529650#comment-529650):_

What traceback are you getting when trying to save?  Saving works for me.  But it fails on load with


```
In [12]: model = datamodels.open("temp.fits")
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~/miniconda3/envs/jwst/lib/python3.9/site-packages/stdatamodels/fits_support.py in from_fits(hdulist, schema, context, skip_fits_update, **kwargs)
    580     try:
--> 581         ff = from_fits_asdf(hdulist, **kwargs)
    582     except Exception as exc:

~/miniconda3/envs/jwst/lib/python3.9/site-packages/stdatamodels/fits_support.py in from_fits_asdf(hdulist, ignore_version_mismatch, ignore_unrecognized_tag, **kwargs)
    608     ignore_missing_extensions = kwargs.pop('ignore_missing_extensions')
--> 609     return fits_embed.AsdfInFits.open(hdulist,
    610                                       ignore_version_mismatch=ignore_version_mismatch,

~/miniconda3/envs/jwst/lib/python3.9/site-packages/asdf/fits_embed.py in open(cls, fd, uri, validate_checksums, extensions, ignore_version_mismatch, ignore_unrecognized_tag, strict_extension_check, ignore_missing_extensions, validate_on_read)
    219         """
--> 220         return cls._open_impl(fd, uri=uri,
    221                        validate_checksums=validate_checksums,

~/miniconda3/envs/jwst/lib/python3.9/site-packages/asdf/fits_embed.py in _open_impl(cls, fd, uri, validate_checksums, extensions, ignore_version_mismatch, ignore_unrecognized_tag, strict_extension_check, _extension_metadata, ignore_missing_extensions, validate_on_read)
    265         try:
--> 266             return cls._open_asdf(self, buff, uri=uri, mode='r',
    267                               validate_checksums=validate_checksums,

~/miniconda3/envs/jwst/lib/python3.9/site-packages/asdf/asdf.py in _open_asdf(cls, self, fd, uri, mode, validate_checksums, do_not_fill_defaults, _get_yaml_content, _force_raw_types, strict_extension_check, ignore_missing_extensions, validate_on_read)
    633         self._fname = self._fd._uri if self._fd._uri else ''
--> 634         header_line = fd.read_until(b'\r?\n', 2, "newline", include=True)
    635         self._file_format_version = cls._parse_header_line(header_line)

~/miniconda3/envs/jwst/lib/python3.9/site-packages/asdf/generic_io.py in read_until(self, delimiter, readahead_bytes, delimiter_name, include, initial_content, exception)
    513         while True:
--> 514             content = reader.read(self.block_size)
    515             buff.write(content)

~/miniconda3/envs/jwst/lib/python3.9/site-packages/asdf/generic_io.py in read(self, nbytes)
    222             if self._exception:
--> 223                 raise ValueError("{0} not found".format(self._delimiter_name))
    224             self._past_end = True

ValueError: newline not found

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
<ipython-input-12-5d7dac46ee31> in <module>
----> 1 model = datamodels.open("temp.fits")

~/dev/jwst/jwst/datamodels/util.py in open(init, memmap, **kwargs)
    192 
    193     # Actually open the model
--> 194     model = new_class(init, **kwargs)
    195 
    196     # Close the hdulist if we opened it

~/dev/jwst/jwst/datamodels/reference.py in __init__(self, init, **kwargs)
     19 
     20     def __init__(self, init=None, **kwargs):
---> 21         super(ReferenceFileModel, self).__init__(init=init, **kwargs)
     22         self._no_asdf_extension = True
     23         self.meta.telescope = "JWST"

~/miniconda3/envs/jwst/lib/python3.9/site-packages/stdatamodels/model_base.py in __init__(self, init, schema, memmap, pass_invalid_values, strict_validation, validate_on_assignment, ignore_missing_extensions, **kwargs)
    212 
    213         elif isinstance(init, fits.HDUList):
--> 214             asdffile = fits_support.from_fits(init, self._schema, self._ctx,
    215                                               **kwargs)
    216 

~/miniconda3/envs/jwst/lib/python3.9/site-packages/stdatamodels/fits_support.py in from_fits(hdulist, schema, context, skip_fits_update, **kwargs)
    581         ff = from_fits_asdf(hdulist, **kwargs)
    582     except Exception as exc:
--> 583         raise exc.__class__("ERROR loading embedded ASDF: " + str(exc)) from exc
    584 
    585     # Determine whether skipping the FITS loading can be done.

ValueError: ERROR loading embedded ASDF: newline not found
```


It looks like this checksum stuff is being passed through `extra_fits`:

```
 'extra_fits.PRIMARY.header.0.0': 'VERIFIED',
 'extra_fits.PRIMARY.header.0.1': 'PASSED',
 'extra_fits.PRIMARY.header.0.2': '',
 'extra_fits.PRIMARY.header.1.0': 'CHECKSUM',
 'extra_fits.PRIMARY.header.1.1': 'NePlQZNjNbNjNZNj',
 'extra_fits.PRIMARY.header.1.2': 'HDU checksum updated 2018-03-07T08:48:27',
 'extra_fits.PRIMARY.header.2.0': 'DATASUM',
 'extra_fits.PRIMARY.header.2.1': '0',
 'extra_fits.PRIMARY.header.2.2': 'data unit checksum updated 2018-03-07T08:48:27',
 'extra_fits.PRIMARY.header.3.0': 'CERTIFYD',
 'extra_fits.PRIMARY.header.3.1': 'PASSED',
 'extra_fits.PRIMARY.header.3.2': '',
 'extra_fits.SCI.header.0.0': 'CHECKSUM',
 'extra_fits.SCI.header.0.1': 'ckaDekTAckYAckYA',
 'extra_fits.SCI.header.0.2': 'HDU checksum updated 2018-03-07T08:48:27',
 'extra_fits.SCI.header.1.0': 'DATASUM',
 'extra_fits.SCI.header.1.1': '843437942',
 'extra_fits.SCI.header.1.2': 'data unit checksum updated 2018-03-07T08:48:27',
 'extra_fits.ASDF.header.0.0': 'CHECKSUM',
 'extra_fits.ASDF.header.0.1': 'S4ALU16JS1AJS15J',
 'extra_fits.ASDF.header.0.2': 'HDU checksum updated 2018-03-07T08:48:27',
 'extra_fits.ASDF.header.1.0': 'DATASUM',
 'extra_fits.ASDF.header.1.1': '3567978628',
 'extra_fits.ASDF.header.1.2': 'data unit checksum updated 2018-03-07T08:48:27'}
```

[Jonathan Eisenhamer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=eisenham) is that desired or expected?  Recall that not all reference files have ASDF extensions, and this is one of them.  I.e.

<https://github.com/spacetelescope/jwst/blob/20d5bbbaeb6c1f25fb23bf12550c01be10c881f7/jwst/datamodels/reference.py#L22>

This seems to be a major bug.

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-2080?focusedCommentId=529659#comment-529659):_

Below is the error I got when trying to save. When creating the files from scratch I didn't get this error, and I use the same statement to add the history string. This is all using v1.2.0.

Traceback (most recent call last):
 File "make_gain_reffiles_using_old_reffiles.py", line 82, in <module>
 model.history.append(history)
 File "/Users/hilbert/miniconda3/envs/mirage-py3.8/lib/python3.8/site-packages/stdatamodels/history.py", line 68, in append
 self._context.add_history_entry(value)
 File "/Users/hilbert/miniconda3/envs/mirage-py3.8/lib/python3.8/site-packages/asdf/asdf.py", line 1242, in add_history_entry
 self.tree['history'].append(entry)
 AttributeError: 'dict' object has no attribute 'append'

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2080?focusedCommentId=529687#comment-529687):_

Also, [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) what was the idea behind not including ASDF extensions in `ReferenceFileModel` which is the base class for so many reference files?  Here's the PR that made that change

#2130

But it's not clear why versioning was a problem.  Is that still a valid reason?

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-2080?focusedCommentId=529693#comment-529693):_

 

IIRC the reasoning for not including the ASDF extension was that for reference files there's no information in the ASDF extension that is not in the FITS file. If the ASDF extension exists it is always checked against changes in the FITS header. Is there a reason why it's needed in reference files?

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-2080?focusedCommentId=529699#comment-529699):_

I don't think this has to do with the attaching the history.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2080?focusedCommentId=529896#comment-529896):_

Agree.

I think the `skip_fits_update` checksumming is not interacting well with datamodels that do not have ASDF extensions.  If one opens these files with `skip_fits_update=False`, things work better.

Unrelated, `util.create_history_entry` now only exists in `stdatamodels`, but we did not change the documentation in `jwst` to indicate this.  And we don't pass it along as an import in `jwst.datamodels.util.create_history_entry`.  We should probably add it or update the docs.

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-2080?focusedCommentId=530019#comment-530019):_

Ugh, never mind. In my original pipeline test script, I ran the pipeline up through the linearity step and saved the results. Then I looped over my collection of new gain reference files and ran ramp-fit with each, starting from the output file from the linearity step. I'm not sure how, but somehow this linearity output file was the cause of the error. 

When I repeated the process above but used the output from the jump step, ramp-fit completed successfully with the new gain reference files.

**Comment by stscijgbot-jp** [comment]
_Comment by [Jonathan Eisenhamer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=eisenham) on [JIRA](https://jira.stsci.edu/browse/JP-2080?focusedCommentId=530020#comment-530020):_

Regardless of the resolution of this particular issue, if the checksum logic does not take into account the fact that there will be no asdf extension on write, that is a bug.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2080?focusedCommentId=532289#comment-532289):_

So coming back to this issue, the original problem of the pipeline crashing while using reffiles produced by the attached `make_gain_reffiles.py` script, i.e.

```
strun calwebb_detector1 jw42424001001_01101_00001_nrca5_uncal.fits --steps.ramp_fit.override_gain=nircam_nrcalong_bootstrapped_gain.fits
```

Turns out not to happen.  User error.  So that is resolved.  It works under all recent versions of the pipeline and its dependencies (`stdatamodels` and `asdf` were checked).

More in a bit on the 2nd issue concerning modifying an existing gain reffile from CRDS that is old enough that it has an ASDF extension.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2080?focusedCommentId=532324#comment-532324):_

Also, the `append(history)` error you're seeing is a bug in the reffile creation script.  If you look at the docs:

<https://jwst-pipeline.readthedocs.io/en/latest/jwst/datamodels/models.html#history-information>

The chunk of history to append needs to be a `History` object.  So if one does

```
from stdatamodels.util import create_history_entry

history = create_history_entry('This file contains...')
model.history.append(history)
```

That solves the issue.  Of course `create_history_entry` is no longer in `jwst.datamodels` now, so we should update our docs to point to it in `stdatamodels`, or better yet, make the `HistoryList.append()` method just automatically wrap whatever is passed to it as a `History` item.  No need to require the user to do all this in a clunky sort of way.  So the API should be improved.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2080?focusedCommentId=532328#comment-532328):_

Finally, the straggler ASDF extension that exists in some of these old reffiles, before we stopped writing them out on subclasses of `ReferenceFileModel` makes the file unreadable after being created.  If they've been created without this flag on read, then the resulting files cannot be opened.  One gets:

```
Traceback (most recent call last):
  File "/Users/jdavies/miniconda3/envs/test/lib/python3.9/site-packages/stdatamodels/fits_support.py", line 581, in from_fits
    ff = from_fits_asdf(hdulist, **kwargs)
  File "/Users/jdavies/miniconda3/envs/test/lib/python3.9/site-packages/stdatamodels/fits_support.py", line 609, in from_fits_asdf
    return fits_embed.AsdfInFits.open(hdulist,
  File "/Users/jdavies/miniconda3/envs/test/lib/python3.9/site-packages/asdf/fits_embed.py", line 218, in open
    return cls._open_impl(fd, uri=uri,
  File "/Users/jdavies/miniconda3/envs/test/lib/python3.9/site-packages/asdf/fits_embed.py", line 262, in _open_impl
    return cls._open_asdf(self, generic_file,
  File "/Users/jdavies/miniconda3/envs/test/lib/python3.9/site-packages/asdf/asdf.py", line 822, in _open_asdf
    header_line = fd.read_until(b'\r?\n', 2, "newline", include=True)
  File "/Users/jdavies/miniconda3/envs/test/lib/python3.9/site-packages/asdf/generic_io.py", line 475, in read_until
    content = reader.read(self.block_size)
  File "/Users/jdavies/miniconda3/envs/test/lib/python3.9/site-packages/asdf/generic_io.py", line 146, in read
    raise DelimiterNotFoundError("{0} not found".format(self._delimiter_name))
asdf.exceptions.DelimiterNotFoundError: newline not found

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/jdavies/miniconda3/envs/test/bin/strun", line 28, in <module>
    step = Step.from_cmdline(sys.argv[1:])
  File "/Users/jdavies/miniconda3/envs/test/lib/python3.9/site-packages/stpipe/step.py", line 173, in from_cmdline
    return cmdline.step_from_cmdline(args)
  File "/Users/jdavies/miniconda3/envs/test/lib/python3.9/site-packages/stpipe/cmdline.py", line 339, in step_from_cmdline
    step.run(*positional)
  File "/Users/jdavies/miniconda3/envs/test/lib/python3.9/site-packages/stpipe/step.py", line 407, in run
    step_result = self.process(*args)
  File "/Users/jdavies/miniconda3/envs/test/lib/python3.9/site-packages/jwst/pipeline/calwebb_detector1.py", line 134, in process
    result, ints_model = self.ramp_fit(result)
  File "/Users/jdavies/miniconda3/envs/test/lib/python3.9/site-packages/stpipe/step.py", line 407, in run
    step_result = self.process(*args)
  File "/Users/jdavies/miniconda3/envs/test/lib/python3.9/site-packages/jwst/ramp_fitting/ramp_fit_step.py", line 53, in process
    gain_model = datamodels.GainModel(gain_filename)
  File "/Users/jdavies/miniconda3/envs/test/lib/python3.9/site-packages/jwst/datamodels/reference.py", line 21, in __init__
    super(ReferenceFileModel, self).__init__(init=init, **kwargs)
  File "/Users/jdavies/miniconda3/envs/test/lib/python3.9/site-packages/stdatamodels/model_base.py", line 232, in __init__
    asdffile = fits_support.from_fits(
  File "/Users/jdavies/miniconda3/envs/test/lib/python3.9/site-packages/stdatamodels/fits_support.py", line 583, in from_fits
    raise exc.__class__("ERROR loading embedded ASDF: " + str(exc)) from exc
asdf.exceptions.DelimiterNotFoundError: ERROR loading embedded ASDF: newline not found
```

I will update AL-556 with info on this bad interaction between old reffiles, `ReferenceFileModel` and `skip_fits_update`.  It's possible a solution is to prevent these keywords from residing in `extra_fits`.  If one opens one of these old reffiles one finds that:

```
>>> print(model.extra_fits.instance)
{'PRIMARY': {'header': [['VERIFIED', 'PASSED', ''], ['CHECKSUM', 'NePlQZNjNbNjNZNj', 'HDU checksum updated 2018-03-07T08:48:27'], ['DATASUM', '0', 'data unit checksum updated 2018-03-07T08:48:27'], ['CERTIFYD', 'PASSED', '']]}, 'SCI': {'header': [['CHECKSUM', 'ckaDekTAckYAckYA', 'HDU checksum updated 2018-03-07T08:48:27'], ['DATASUM', '843437942', 'data unit checksum updated 2018-03-07T08:48:27']]}, 'ASDF': {'header': [['CHECKSUM', 'S4ALU16JS1AJS15J', 'HDU checksum updated 2018-03-07T08:48:27'], ['DATASUM', '3567978628', 'data unit checksum updated 2018-03-07T08:48:27']]}}
```

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2080?focusedCommentId=532334#comment-532334):_

Finally, if opening one of the existing, old reffiles which has an old-type ImageHDU ASDF extension, and using it to make a new on as in the attached script above, if one puts a

```
model.extra_fits = {}
```

into the script before saving, they will write out and be able to be read in and used in the pipeline without error.  So this hints at what `stdatamodels` should be doing in terms of fixing this behavior.

Might not be a bad idea to toss a

```
with datamodels.open(outfile) as model:
    assert model.meta.model_type == "GainModel"
```

at the end of the reffile creation script as a sanity check, to verify it loads and verify anything else you might want to check about the file.

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2080?focusedCommentId=533424#comment-533424):_

This is now fixed in stdatamodels 0.2.3 release.
